### PR TITLE
SCI32: Fix QFG4 tarot readings

### PR DESCRIPTION
**Pull request by Vhati**
Fixes incorrect cards and placement in the third and fifth readings.

[#10822](https://bugs.scummvm.org/ticket/10822)
```
// The fortune teller's third reading places the "Two of Cups" across another
// card, off-center. That View (1023) is cropped (130x64). All other horizontal
// cards (130x86, 130x87) are padded at the bottom with transparent pixels.
//
// A utility script (sShowCard) is scheduled to create each card with a given
// View (passed as local[0]) and move it onto a given pile (passed as local[2]:
// center, west, south, east, north). It has a switch block deciding x,y coords
// depending on the pile.
//
// We optimize a couple cases by consolidating their common code in a
// subroutine to make room. The rewritten case for the east pile checks if the
// requested card was "Two of Cups". If so, a special Y value is used.
```
[#10823](https://bugs.scummvm.org/ticket/10823)
```
// The fortune teller's third reading displays a right-turned "The Devil" card,
// but Magda says it is "Death".
//
// We change the card to a right-turned "Death".
```
[#10824](https://bugs.scummvm.org/ticket/10824)
```
// The fortune teller's third reading has the wrong card at the center. The 3rd
// and 4th reading are about different people, yet both have "Queen of Cups".
//
// The 1st reading establishes "Queen of Cups" as: a woman of wisdom and love,
// kind, generous, and virtuous. This fits the 4th reading: she uses her power
// joyfully, giving gracefully and lovingly to others.
//
// The 1st reading establishes "Queen of Swords" as: a deceiver or deceived,
// having suffered through terrible hardship, she faces her sorrows bravely,
// but with deep loneliness. This fits the 3rd reading better: some cruel event
// shaped her life... ambition, self-deception, and she is falling in love.
//
// We change the 3rd reading's center card to "Queen of Swords".
```
[#10845](https://bugs.scummvm.org/ticket/10845)
````
// The fortune teller's third reading places horizontal cards on top of
// vertical ones. Some then fall *through* the vertical card to the bottom.
//
// A utility script (sShowCard) creates each card and moves it onto a pile
// (center, west, south, east, north). Then the card is assigned a priority
// with setPri(). Generally these piles are two cards deep. Center has one.
//
// Every pile ought to start with priority 0 and increment thereafter. Somebody
// mixed up the setPri() sequence. We change 0;1,0;1,0 to 0;0,1;0,1.
````
[#10846](https://bugs.scummvm.org/ticket/10846)
````
// The fortune teller's fifth reading is unusual. It places all cards at the
// center pile, periodically fading out to clear the table. When Magda
// talks about the Sense Ritual, the "Six of Swords" (view 1048) sinks below
// the previous card.
//
// A shared utility script that creates and places cards (sSetTheSignificator)
// uses priority 12 as it deals, after which each card is given a lower value
// with setPri(). "The Falling Tower" (view 1031) did *not* get a new priority.
// Thus "Six of Swords", when given priority 1, sinks below 12.
//
// "Six of Swords" is the last card before a fade. We simply leave its priority
// at 12 as well. Being the most recent card, it will be on top. No worry
// about covering a subsequent card because the table will be cleared.
````

To test any reading...
* Create a new character. Any kind.
* Set plot flags for having seen previous readings.
  * Reading 1.
    * Just teleport.
  * Reading 2 (297).
    * vv g 518 64
  * Reading 3 (297, 298).
    * vv g 518 96
  * Reading 4 (297, 298, 299).
    * vv g 518 112
  * Reading 5 (297, 298, 299, 300).
    * vv g 518 120
* Teleport into the tarot room.
  * room 475

For anyone interested in seeing higher-res cards...

[Video](https://www.youtube.com/watch?v=7SD0pAkPkRM): YouTube - Spoiler of the "Russian Tarot of St. Petersburg" deck (illus. Yury Shakov)
* 0:31 Death
* 0:34 The Devil
* 0:46 Queen of Swords
* 1:14 Queen of Cups
* 1:25 Two of Cups

### CODE REVIEW

File: engines/sci/engine/script_patches.cpp [hunk 1]
```diff
@@ -8579,6 +8581,150 @@ static const uint16 qfg4DoubleDoorSoundPatch[] = {
...
+
+// The fortune teller's third reading places the "Two of Cups" across another
+// card, off-center. That View (1023) is cropped (130x64). All other horizontal
+// cards (130x86, 130x87) are padded at the bottom with transparent pixels.
+//
+// We optimize the method that creates and places cards, then specifically give
+// this card a fudged y position. Our subroutine reduces the size of cases in a
+// switch. We could rewrite all its cases, but just rewriting case 4 and 5 is
+// enough.
+//
```
Comment (bluegr) [inline engines/sci/engine/script_patches.cpp]: Can't you just change the card's y position, instead of rewriting the whole method?

Comment (Vhati) [reply engines/sci/engine/script_patches.cpp]: For each card, a shared method is responsible for creating the card (with the given view) and giving it a motion to land on one of several piles (the desired pile number is also given). There are north, south, east, west, and center piles. Multiple cards are moved onto each pile.

I can't give a specific position before the shared method. There's no card. And it IS going to the proper pile, except that that view has weird dimensions, so the method's normal pile coords are inappropriate. If I repositioned the card after the shared method, the card would jump or move twice.

If I changed the hardcoded coords the shared method uses for the east pile, that would affect all cards that land there, not just the weird card.

**Comment by Vhati** [comment]
This is what sThirdReading::changeState() looks like. It's basically a cutscene, where the player clicks to deal the the next card and hear Magda's interpretation.

````
[...]
(18
	(= local0 1022)
	(= local2 4)
	(g0_hero setScript: sShowCard self)
)
(19
	(local1 setPri: 1)
	(g91_gloryMessager say: 1 6 36 0 self)  # Magda describes the card.
)
(20 (localproc_1cfb))
(21
	(= local0 1023)
	(= local2 4)
	(global0 setScript: sShowCard self)
)
(22
	(local1 setPri: 2)
	(g91_gloryMessager say: 1 6 37 0 self)  # Magda describes the card.
)
(23 (localproc_1cfb))
(24
	(= local0 1024)
	(= local2 5)
	(g0_hero setScript: sShowCard self)
)
[...]
````
local[0] is the desired view for the card.
local[2] is the numbered destination pile.
sShowCard creates and places the card.
local[1] is the last created card.

localproc_1cfb() selects HAND on g69_mainIconBar and calls g80_User::canInput(1), allowing the player to click the deck for another card.

**Comment by Vhati** [comment]
script 475 - sShowCard::changeState()
````
(switch (= state param1)
	(0
		(global1 handsOff:)
		((= local1 (aTurnCard new:))
			view: local0
			init:
			moveSpeed: 0
			setPri: 250
			setStep: 30 10
			setMotion: MoveTo 148 150 self
		)
	)
	(1 (= seconds 5))
	(2
		(switch local2
			(1  # Center pile
				(local1
					moveSpeed: 0
					setStep: 30 10
					setScaler: Scaler 100 35 150 110
					setMotion: MoveTo 145 110 self
				)
			)
			(2  # West pile
				(local1
					moveSpeed: 0
					setStep: 30 10
					setScaler: Scaler 100 35 150 110
					setMotion: MoveTo 80 110 self
				)
			)
			(3  # South pile
				(local1
					moveSpeed: 0
					setStep: 30 10
					setScaler: Scaler 35 100 170 150
					setMotion: MoveTo 144 170 self
				)
			)
			(4  # East pile
				(local1
					moveSpeed: 0
					setStep: 30 10
					setScaler: Scaler 100 35 150 110
					setMotion: MoveTo 210 110 self

# This is where Two of Cups gets a fudged Y: 102 instead of 110.
				)
			)
			(5  # North pile
				(local1
					moveSpeed: 0
					setStep: 30 10
					setScaler: Scaler 100 35 150 50
					setMotion: MoveTo 144 50 self
				)
			)
	(3 (sndPlace play: self))
	(4
		(= local2 0)
		(self dispose:)
	)
)
````

**Comment by Vhati** [comment]
Description has been revised.

**Comment by Vhati** [comment]
Added two more patches, for priority bugs in the third and fifth readings.

**Comment by Vhati** [comment]
I broke the "off-center Two of Cups" patch just now in an attempt to shave off a few bytes. Working on it...

**Comment by Vhati** [comment]
While I caught up with master, I'd trimmed some bytes switching call offsets from UINT16 to single-byte. Waste bytes' disassembly happened to align without the need for an erasure line.

Forgot to adjust jumps and branches afterward. Should be working again now.

**Comment by bluegr** [comment]
Nice work! This is quite a long list of bugs for the card readings, and your time spent on these is much appreciated! :)

Bugs [#10822](https://bugs.scummvm.org/ticket/10822), [#10845](https://bugs.scummvm.org/ticket/10845) and [#10846](https://bugs.scummvm.org/ticket/10846) are all minor graphical glitches.

Bug [#10823](https://bugs.scummvm.org/ticket/10823) is obviously wrong, the voice describes a different card.

Bug [#10824](https://bugs.scummvm.org/ticket/10824) is quite obscure. After reading through all the transcripts as you wrote them in the bug, it's quite apparent that the central card is wrong in the third reading.

Patching these (especially [#10822](https://bugs.scummvm.org/ticket/10822)) was hard. Kudos for your work!

Merging

**Comment by Vhati** [comment]
@bluegr:
> your time spent on these is much appreciated! :)

I appreciate your closing remarks on tickets and PRs.
Even the boilerplate ones are good for morale.

**Pull request merged**